### ISSUE TITLE
A PKCS#8 certificate incorrectly formatted as a PKCS#1 certificate breaks haproxy

### ISSUE THREAD

**Issue by jmakin-godaddy** [haproxy/haproxy#2166]
### Detailed Description of the Problem

If the private key of a PKCS#8 certificate in PEM form, which should begin and end with:
BEGIN PRIVATE KEY
and
END PRIVATE KEY
respectively, are accidentally posing as a PKCS#1 certificate, by beginning and ending with:
BEGIN RSA PRIVATE KEY
and
END RSA PRIVATE KEY
the config check will fail and the daemon will not start, and the error message provided is misleading: "unable to load certificate chain from file" - the problem in this particular scenario isn't to do with the chain

### Expected Behavior

The config check could pass, as the certificate can still be loaded (with a tweak listed below)

### Steps to Reproduce the Behavior

Have a front-end configured with a crt-list of pem format files
Edit a pem, swapping out PRIVATE KEY for RSA PRIVATE KEY
Either attempt to start the daemon or do a config check: haproxy -c -f /path/to/config

### Do you have any idea what may have caused this?

Why I have certificates in the wrong format? Not sure
Why the code assumes failing at that point is a chain problem? Not sure

### Do you have an idea how to solve the issue?

The code involved is in src/ssl_ckch.c

I have been running version 2.2 and have successfully fixed the problem with the following change:
```
--- haproxy-2.2.29/src/ssl_ckch.c	2023-02-14 15:57:38.000000000 +0000
+++ haproxy-2.2.29/src/ssl_ckch2.c	2023-05-17 18:23:46.544704610 +0100
@@ -493,9 +493,9 @@
 	}
 
 	ret = ERR_get_error();
-	if (ret && (ERR_GET_LIB(ret) != ERR_LIB_PEM && ERR_GET_REASON(ret) != PEM_R_NO_START_LINE)) {
-		memprintf(err, "%sunable to load certificate chain from file '%s'.\n",
-		          err && *err ? *err : "", path);
+	if (ret && (ERR_GET_LIB(ret) != ERR_LIB_PEM && ERR_GET_REASON(ret) != PEM_R_NO_START_LINE && ERR_GET_REASON(ret) != ASN1_R_WRONG_TAG)) {
+		memprintf(err, "%sunable to load certificate chain from file '%s' : '%d' - '%d' - '%s'.\n",
+		          err && *err ? *err : "", path, ERR_GET_REASON(ret), ERR_GET_LIB(ret), ERR_lib_error_string(ret));
 		goto end;
 	}
 ```

This will also work for haproxy version 2.4, but the latest version is slightly changed

### What is your configuration?

```haproxy
REDACTED
```

### Output of `haproxy -vv`

```plain
$ haproxy -vv                                                                                                                                                                                                  [6/49941]
HA-Proxy version 2.2.29-c5b927c 2023/02/14 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.29.html
Running on: Linux 3.10.0-1160.62.1.el7.x86_64 #1 SMP Tue Apr 5 16:57:59 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL +EPOLL -EVPORTS +FUTEX +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY -LUA +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL +PCRE -PCRE2 
-PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PRIVATE_CACHE -PTHREAD_PSHARED +RT -SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 4.8.5 20150623 (Red Hat 4.8.5-44)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace
        [CACHE] cache
        [FCGI] fcgi-app
```

### Additional Information

My operating system for this is centos7 but I believe this issue is independent of that.
Openssl version 1.0.2k

**Comment by wlallemand** [comment]
Since v2.7 HAProxy is a little bit more verbose about loading errors. We introduced the ERR_reason_error_string() function in a lot of code path (ex f784b90eae4c5f27fd8fa4a4ba2c72a03993ae94). 

If your certificate is invalid, HAProxy must not start, we are stricter in newer versions of HAProxy on the loading errors (78c7a06e4fc1214b40845e6a61073e558fd6a02b). This was not backported since it could break a lot a setup during a minor update.

In my opinion a loading error must happen here, but the ERR_reason_error_string() function is needed so people can fix their certificates. But we can't backport the fix about loading unfortunately.

**Comment by wlallemand** [comment]
Also I couldn't reproduce your problem by switching to  BEGIN RSA PRIVATE KEY, something else should be wrong, but we definitively don't want to ignore the error.

**Comment by wlallemand** [comment]
It looks like I didn't analyzed this correctly, I can reproduce your problem in 2.5.0 but not in 2.6.0, I'm still looking for the commit that changed it, it could have been fixed by accident.

**Comment by wlallemand** [comment]
Looks like it was broken by accident in c76c3c4e59c8c0b46ef15544bdc03afcad678b22, which does a ERR_clear_error() after the PEM_read_bio_PrivateKey and PEM_read_bio_DHparams, previously the error was kept and returned too late, letting the user think this is a chain error. But in fact this is a PEM which happens at the certificate chain step, but if I move the error above it would happen at the DH step...

`[ALERT]    (184625) : config : parsing [haproxy.cfg:45] : 'bind *:8443' : unable to load DH from file 'foo.pem': asn1 encoding routines::nested asn1 error.`

So in conclusion you MUST have this error because your file is invalid, and openssl returns an error.
